### PR TITLE
improve and consolidate upgrade path decision logic

### PR DESCRIPTION
**Pull request by mambocab**
This rewrites the logic for determining the upgrade path in the storage engine upgrade tests. Rather than working off the branch name, it works off the version of the It works locally for me, both for running as-is, and for specifying paths with `UPGRADE_TO`. Demonstration of the latter is here:

http://cassci.datastax.com/job/scratch_mambocab-storage_engine_upgrades-2.1/17

This should deal with problems like [the problem tjake had](https://issues.apache.org/jira/browse/CASSANDRA-10261?focusedCommentId=14740734&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14740734) where the test ran the upgrade from cassandra-3.0 to the pre-3.0 dev branch, which isn't a supported path.

### CODE REVIEW

File: upgrade_tests/upgrade_base.py [hunk 1]
```diff
@@ -25,6 +26,39 @@
...
+        upgrade_version = 'git:cassandra-2.2'
+    elif '3.0' <= job_version < '3.1':
+        # We can choose 2.2 because it will run on JDK 1.8
+        start_version = 'git:cassandra-2.2'
+    elif '3.1' <= job_version:
+        # 2.2->3.X, where X > 0, isn't a supported upgrade path,
+        # but 3.0->3.X is.
+        start_version = 'git:cassandra-3.0'
+
+    assert [start_version, upgrade_version].count(None) >= 1
```
Comment (ptnapoleon) [inline upgrade_tests/upgrade_base.py]: Why not use a better assert?

**Comment by ptnapoleon** [comment]
If the **install_kwargs logic works, and you add an assert message, I'm +1

**Comment by mambocab** [comment]
I've added an assert message. The install kwargs logic does work -- I'll run the dtests on trunk and 3.0 with this dtest revision here:

http://cassci.datastax.com/view/All_Jobs/job/scratch_mambocab-storage_engine_upgrades-trunk/

and here:

http://cassci.datastax.com/view/All_Jobs/job/scratch_mambocab-storage_engine_upgrades-3.0/

And merge if that confirms it works.

**Comment by mambocab** [comment]
Also waiting on a confirmation job here:

http://cassci.datastax.com/view/All_Jobs/job/scratch_mambocab-dtest_windows/

**Comment by mambocab** [comment]
@ptnapoleon I'm getting errors like this:

http://cassci.datastax.com/view/All_Jobs/job/scratch_mambocab-dtest_windows/1/testReport/upgrade_tests.cql_tests/TestCQL/blobAs_functions_test/

on the Windows job. The other two jobs are reasonable:

http://cassci.datastax.com/view/All_Jobs/job/scratch_mambocab-storage_engine_upgrades-3.0/lastCompletedBuild/testReport/

http://cassci.datastax.com/view/All_Jobs/job/scratch_mambocab-storage_engine_upgrades-trunk/lastCompletedBuild/

Would it be all right to add a `@skipIf(sys.platform == 'win32', 'Skip upgrade tests on Windows')` for now and merge?

**Comment by ptnapoleon** [comment]
Yes, it would. +1

**Comment by mambocab** [comment]
Ok, stuck some band-aids on it. @ptnapoleon can you have one last look before I merge?

**Comment by ptnapoleon** [comment]
+1 to the latest commit

**Comment by mambocab** [comment]
Thank you -- I'm running those confirmation jobs again, and unless something terrible happens, I'll merge.

**Comment by mambocab** [comment]
Ok, this has run successfully here:

http://cassci.datastax.com/view/All_Jobs/job/scratch_mambocab-trunk-check/1/testReport/

Merging.

**Pull request merged**